File: backend/schemas/research_explorer.py
from pydantic import BaseModel

from schemas.linking import LinkableObject, LinkableType


class ExplorerSearchResponse(BaseModel):
    objects: list[LinkableObject]


class CorpusEntry(BaseModel):
    object: LinkableObject
    roles: list[str] = []


class ExplorerObjectDetail(BaseModel):
    object: LinkableObject
    description: str = ""
    parent: str = ""
    tags: list[str] = []
    metadata: dict[str, str] = {}
    related: dict[LinkableType, list[LinkableObject]]
    backlinks: dict[LinkableType, list[LinkableObject]]
    all_related: list[LinkableObject]
    corpus: dict[LinkableType, list[CorpusEntry]] = {}

File: frontend/src/lib/features/research-explorer/types/researchExplorer.ts
import type {
  LinkableObject,
  LinkableType
} from '$lib/features/linking/types/linking';

export type ExplorerObjectDetail = {
  object: LinkableObject;
  description: string;
  parent: string;
  tags: string[];
  metadata: Record<string, string>;
  related: Record<LinkableType, LinkableObject[]>;
  backlinks: Record<LinkableType, LinkableObject[]>;
  all_related: LinkableObject[];
  corpus?: Partial<Record<LinkableType, CorpusEntry[]>>;
};

export type ExplorerSearchResponse = {
  objects: LinkableObject[];
};

export type CorpusEntry = {
  object: LinkableObject;
  roles: string[];
};
